### ISSUE TITLE
-retry option seems not working

### ISSUE THREAD

**Issue by PROBIOTICS4** [tnpitsecurity/ligolo-ng#9]
Thank you for the effort to bring this great tool. When I'm explorering its capability, I found out that the `-retry` option seems not working, here is my experiment env:
1.vps aka proxy (win10)
2.my pc aka agent (win10)

**step 1.** vps start commandline:
```
proxy.exe -laddr 0.0.0.0:22 -selfcert -tun tun0 -v
```
Proxy output:
```
C:\Users\Administrator\Desktop\ligolo-ng_proxy_0.3_Windows_64bit>proxy.exe -laddr 0.0.0.0:22 -selfcert -tun tun0 -v
2021/12/17 15:17:48 Using existing driver 0.14
2021/12/17 15:17:48 Creating adapter
time="2021-12-17T15:17:48+08:00" level=warning msg="Using automatically generated self-signed certificates (Not recommended)" func="ligolo-ng/pkg/proxy.(*Controller).ListenAndServe" file="/home/runner/work/ligolo-ng/ligolo-ng/pkg/proxy/controller.go:65"
time="2021-12-17T15:17:48+08:00" level=info msg="Listening on 0.0.0.0:22" func="ligolo-ng/pkg/proxy.(*Controller).ListenAndServe" file="/home/runner/work/ligolo-ng/ligolo-ng/pkg/proxy/controller.go:135"
2021/12/17 15:17:48 Removed orphaned adapter "tun0 1"
    __    _             __
   / /   (_)___ _____  / /___        ____  ____ _
  / /   / / __ `/ __ \/ / __ \______/ __ \/ __ `/
 / /___/ / /_/ / /_/ / / /_/ /_____/ / / / /_/ /
/_____/_/\__, /\____/_/\____/     /_/ /_/\__, /
        /____/                          /____/

Made in France ♥ by Cha! - TNP IT Security <tnpitsecurity.com>

```
**step 2.** agent start commandline:
```
agent.exe -connect vps_ip:22 -ignore-cert -retry -v
```
Proxy output (Agent joined):
```
ligolo-ng » time="2021-12-17T15:18:08+08:00" level=info msg="Agent joined." func=main.main.func1 file="/home/runner/work/ligolo-ng/ligolo-ng/cmd/proxy/main.go:70" name="WIN-xxxxxxxxxxxxxxU\\Administrator@WIN-xxxxxxxxxxxxxxU" remote="_some_ip_:37134"
ligolo-ng »
ligolo-ng »
```

**step 3.** shutdown proxy by manually Ctrl-C and restart proxy by the same command
Proxy output (input Ctrl-c and restart):
```
ligolo-ng »
ligolo-ng »
ligolo-ng » ^C
input Ctrl-c once more to exit
ligolo-ng » ^C
interrupted

C:\Users\Administrator\Desktop\ligolo-ng_proxy_0.3_Windows_64bit>proxy.exe -laddr 0.0.0.0:22 -selfcert -tun tun0 -v
2021/12/17 15:18:13 Using existing driver 0.14
2021/12/17 15:18:13 Creating adapter
time="2021-12-17T15:18:14+08:00" level=warning msg="Using automatically generated self-signed certificates (Not recommended)" func="ligolo-ng/pkg/proxy.(*Controller).ListenAndServe" file="/home/runner/work/ligolo-ng/ligolo-ng/pkg/proxy/controller.go:65"
time="2021-12-17T15:18:14+08:00" level=info msg="Listening on 0.0.0.0:22" func="ligolo-ng/pkg/proxy.(*Controller).ListenAndServe" file="/home/runner/work/ligolo-ng/ligolo-ng/pkg/proxy/controller.go:135"
2021/12/17 15:18:14 Removed orphaned adapter "tun0 1"
    __    _             __
   / /   (_)___ _____  / /___        ____  ____ _
  / /   / / __ `/ __ \/ / __ \______/ __ \/ __ `/
 / /___/ / /_/ / /_/ / / /_/ /_____/ / / / /_/ /
/_____/_/\__, /\____/_/\____/     /_/ /_/\__, /
        /____/                          /____/

Made in France ♥ by Cha! - TNP IT Security <tnpitsecurity.com>

ligolo-ng »
```
**step 4.** Agent retry connecting proxy but failed.
Agent output(retry failed):
```
time="2021-12-17T15:21:12+08:00" level=warning msg="warning, certificate validation disabled" func=main.main file="ligolo-ng/cmd/agent/main.go:82"
time="2021-12-17T15:21:12+08:00" level=info msg="Connection established" func=main.connect file="ligolo-ng/cmd/agent/main.go:110" addr="vps_ip:22"
222222222
time="2021-12-17T15:21:17+08:00" level=error msg="Connection error: read tcp 192.168.121.130:49188->vps_ip:22: wsarecv: An existing connection was forcibly closed by the remote host." func=main.main file="ligolo-ng/cmd/agent/main.go:90"
time="2021-12-17T15:21:17+08:00" level=info msg="Retrying in 5 seconds." func=main.main file="ligolo-ng/cmd/agent/main.go:92"
time="2021-12-17T15:21:22+08:00" level=info msg="Connection established" func=main.connect file="ligolo-ng/cmd/agent/main.go:110" addr="vps_ip:22"
222222222
time="2021-12-17T15:21:22+08:00" level=error msg="Connection error: write tcp 192.168.121.130:49188->vps_ip:22: use of closed network connection" func=main.main file="ligolo-ng/cmd/agent/main.go:90"
time="2021-12-17T15:21:22+08:00" level=info msg="Retrying in 5 seconds." func=main.main file="ligolo-ng/cmd/agent/main.go:92"
```

It seems agent can not connect proxy by using `-retry` options, I put fmt.Println("2222222") in `cmd\agent\main.go:115`. I think the problem is located at followed code in `cmd\agent\main.go:101-120`:
```
func connect(conn net.Conn, config *tls.Config) error {
	tlsConn := tls.Client(conn, config)

	yamuxConn, err := yamux.Server(tlsConn, yamux.DefaultConfig())
	if err != nil {
		fmt.Println("1111111111")
		return err
	}

	logrus.WithFields(logrus.Fields{"addr": tlsConn.RemoteAddr()}).Info("Connection established")

	for {
		conn, err := yamuxConn.Accept()
		if err != nil {
			fmt.Println("222222222")
			return err
		}
		go handleConn(conn)
	}
}
```
There is something wrong with `yamux` dealing with `conn`.
FYI, I used Wireshark try to capture stream but ,surprisingly ,agent didn't send TCP package during retrying.
Please check if you have the same issue, if does, please fix it, thank you!

**Comment by Nicocha30** [comment]
Thanks for the bug report :tada: , fixed in the latest release (v0.3.1).

**Comment by PROBIOTICS4** [comment]
🎉🎉🎉🎉Thank you!

> Thanks for the bug report 🎉 , fixed in the latest release (v0.3.1).
